Fixed issue with collapsable ribbon and quick acces toolbar

APEbbers opened the pull request:
Hi,

I went a bit futher with the ribbon. I noticed that when you collapse the ribbon, the area didn't shrink with it. Also when I opened FreeCAD on my laptop screen, the quickaccess toolbar dissapeard.

I totally understand if I overstept any boundary, so feel free to ignore all below. I can imagen that you have other plans with this workbench.

The biggest change is that I removed the customized qtpy folder.
After a few attempts, I was stuck and lost a bit of track with all the changes that I had done. I was curious to see what happend if i use the original qtpy package. As it turns out, the ribbon collapsed as it should. So I went a bit furtner with the changes below.

I've added the following:
-  the width of the quick access toolbar is now based on the number of buttons.
- the height of the quick access toolbar and the application button are tweaked a little bit.
- I Moved the "geticonsize" to the class level. This way you can use it with all defs in the class
- the tabs are now sorted based on the order of the workbenches as defined in preferences->workbenches
- I've added you customized stylesheets and have the qtpy defaults overridden with it.
-  I've added qtpy and pyqtribbon as a dependeny in package.xml. All customizations are now done in FCBinding.py. 

kind regards

Paul

geolta commented:
Hi, these look like nice quality of life improvements :)
Unfortunately I'm currently in vacation without my laptop and therefore not able to properly review your changes. But I promise you to look at them when I'm back (in ~3 weeks).
Keep your improvements flowing, they are pushing this repository from a testing state to something usable, maybe we can then even add it to the Addon Manager :)
Thank you again!

APEbbers commented:
Hi,

a short update about my progress. I've created a settings dialog which you can open via a button under the application menu.
With this dialog, you can change the ribbon design. you can change the ribbon buttons per workbench and then per toolbar.
To achieve this, I've added two levels above "Toolbars" -> "workbenches" & "workbench name".
I now people prefer a drag/drop function, but I figured that with this dialog in place, we have more time to look into that.

When you open the settings dialog for the first time, all workbenches will be loaded. I noticed that when you don't activate a workbench, you do not get its toolbars and commands properly. So it takes a little bit of time to open at the moment.

Overall I changed/added the following:
- Added a helpbutton with the link to the help as set in the FreeCAD preferences
- Added a settings button under the application button to open the settings dialog
- Added a separate file called "Parameters_Ribbon.py". I used this to centralize all settings that we can use in a preference page.
- Added a file with standard functions. (Is a collection of functions that I created for my workbenches)
- Added a settings dialog with the following functions:
   - Select quick access commands
   - In/Exclude workbenches
   - In/Exclude toolbars
   - Set the Icon size for commands per workbench, per toolbar
   - Buttons to change the order of commands
   - All Icons are now from FreeCAD itself including dropdown buttons (I took the first command in the dropdown)
   - Restore function.
   - Reset function. (a proper default layout has to be made yet)
   - Settings tab with the folowing settings:
       - Enable/Disable the backup function
       - Set the backup location
       - Enable or disable AutoHide
       - Change the size of the small, medium and large buttons
       - Load a stylesheet 

To generate a default layout for the reset button I'm thinking to create a separate script, that set every first command in a toolbar to large. What is your opinion about this? Or we make a default for a few internal workbenches and the rest is up to the users?

I didn't started with a preference page yet. I've made two workbenches so far and honestly I don't like the preference page that much. It is limited in what you can do with it.

kind regards

APEbbers commented:
I've added a script to generate a json file. It is standalone, so you can run it directly in FreeCAD.
The script set every first command of a toolbar to large. 
I believe it helps to create a default for the ribbon.

The next thing I'd like to do is add the possibility to combine toolbars to one big toolbar.
I think if we have that, we have a good base to release the ribbon as an add-on. And develop further from there.

APEbbers commented:
You can now sort the panels, create custom panels and custom toolbars are now included

APEbbers committed:
Merge pull request #25 from APEbbers/Develop

- Fixed issue where the command on multiple toolbars, shows the checkstate from the first toolbar
- Update button will be activated only when there is a difference between the settings and the json file
- Added a submenu with scripts to help setup the ribbon
- Moved custom scripts to dedicated folder
- Apdated the resetscript and renamed it to "CreateDefaultRibbonStructure"

APEbbers committed:
- Fixed issue where the toolbars/panel order was reverted back when opening the design dialog
- Fixed issue where the json file overwrite changes when the dialog was opened a second time

APEbbers committed:
increase version

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #26 from APEbbers/Develop

Develop

APEbbers committed:
Revert "Develop"

APEbbers committed:
Merge pull request #27 from APEbbers/revert-26-Develop

Revert "Develop"

APEbbers committed:
- Fixed issue where the toolbars/panel order was reverted back when opening the design dialog
- Fixed issue where the json file overwrite changes when the dialog was opened a second time

APEbbers committed:
Merge branch 'main' into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #28 from APEbbers/Develop

- Fixed issue where the toolbars/panel order was reverted back when o…

APEbbers committed:
Merging gone wrong

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #29 from APEbbers/Develop

Merging gone wrong

APEbbers committed:
Small updates for FCBinding and LoadDesign_Ribbon
Added GenerateCommandsList.py and UpdateCommandNames.py

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

APEbbers committed:
Updated initGui.py to prepare for AddOn
implemented addAction function for the ribbonbar. This makes the ribbon compatible with other add-ons that use a accessiores menu

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #30 from APEbbers/Develop

implemented accessories menu

APEbbers committed:
Updated package.xml -> no compatibility issues anymore

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

APEbbers committed:
Added a pin button to the right toolbar.
The settings in the preference menu are now stored in parameters instead of the Json file.
Json functions are removed from the preference menu

APEbbers committed:
reverted initGui.py back
updated pin icon

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #31 from APEbbers/Develop

Develop

APEbbers committed:
Fixed issue with checkchanges still looking in json file for showtext lines

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #32 from APEbbers/Develop

Fixed issue with checkchanges still looking in json file for showtext…

APEbbers committed:
Added the Ribbon design button and scripts to one submenu

APEbbers committed:
Fixed issue where close button was not working for the preference menu

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #33 from APEbbers/Develop

Fixed issue with close button not working for preference menu

APEbbers committed:
Start implentation of separators in panels

APEbbers committed:
Separators can now be added to panels.
Icons for buttons to be done

APEbbers committed:
Update FCBinding.py

APEbbers committed:
Updated design UI + remove separator now works

APEbbers committed:
Merge branch 'main' into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Add/Remove buttons changed to pushbuttons for proper representation
RibbonStructure.json updated with separators for PartDesign and FEM toolbars

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Fixed issue in settings menu

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #34 from APEbbers/Develop

Added possibility to add separators to panels

APEbbers committed:
tweaked toolbar buttons to avoid buttons disapering by certain freecad themse

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #35 from APEbbers/Develop

tweaked toolbar buttons to avoid buttons disapering by certain freeca…

APEbbers committed:
Fixed issue where the icononly settings was ignored

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #36 from APEbbers/Develop

Develop

APEbbers committed:
changed dropdown buttons to menubutton popup + changed the minimum width for small and medium icons

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #37 from APEbbers/Develop

Develop

geolta commented:
Hi, thank you again for your work and sorry for my late reply :)

The problem is, that this PR does a lot of things at once and has 15k lines of code added (although most of them are the JSON file and the QtDesigner UI files). This is not reviewable for me. Is it feasible for you to split this into multiple smaller PRs? I totally understand if this is too much work for you. Then I would limit my review to thorough testing and a few important code snippets :)

Best regards!

APEbbers commented:
Hi,

 

I don’t think it is possible to split up the pull request. But maybe you can create a second branch? I then create a pull request to that one, so you can review it without changing the main branch?

 

Best regards!

 

From: geolta ***@***.***> 
Sent: Friday, 6 September 2024 12:11
To: geolta/FreeCAD-Ribbon ***@***.***>
Cc: Paul Ebbers ***@***.***>; Author ***@***.***>
Subject: Re: [geolta/FreeCAD-Ribbon] Fixed issue with collapsable ribbon and quick acces toolbar (PR #13)

 

Hi, thank you again for your work and sorry for my late reply :)

The problem is, that this PR does a lot of things at once and has 15k lines of code added (although most of them are the JSON file and the QtDesigner UI files). This is not reviewable for me. Is it feasible for you to split this into multiple smaller PRs? I totally understand if this is too much work for you. Then I would limit my review to thorough testing and a few important code snippets :)

Best regards!

—
Reply to this email directly, view it on GitHub <https://github.com/geolta/FreeCAD-Ribbon/pull/13#issuecomment-2333727941> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACNM6X3VC4LFUACTYM7EGKTZVF5RNAVCNFSM6AAAAABLW7IMQ6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMZTG4ZDOOJUGE> .
You are receiving this because you authored the thread.  <https://github.com/notifications/beacon/ACNM6XY5UQUTME4CMA4WPNDZVF5RNA5CNFSM6AAAAABLW7IMQ6WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTULDHOMK.gif> Message ID: < ***@***.***> ***@***.***>

APEbbers committed:
Fixed issue with freecad preference page crashing on ok click
This was caused by wrong implentation of the ribbon into the main window.
Collapse ribbon and hover behavior had to be rewritten

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #38 from APEbbers/Develop

Fixed issue with freecad prefenence page crashing. This was caused by wrong implentation of the ribbon into the main window.
Now the ribbon is added as a dockwidget (as already done by Geolta). Functions for autohide, and hover are added to bring back the proper behavior hiding the ribbon

APEbbers committed:
Fixed issue where initial settings where not set. This resulted in too small buttons

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #39 from APEbbers/Develop

Fixed issue where initial settings where not set. This resulted in to…

APEbbers committed:
Fixed issue where large icons are always showing text

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

APEbbers committed:
Fixed issue where dropdown arrow on large toolbuttons was infront of icon

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #40 from APEbbers/Develop

Develop

APEbbers committed:
Added the menus of the standard menubar to the application button
With the alt key you can show/hide the standard menubar
added keybord to the dependencies

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #41 from APEbbers/Develop

Added the menus of the standard menubar to the application button

APEbbers commented:
Hi,

Did you had a look at my updates? Fixed a few issues. The way I loaded the ribbon, caused the freecad preferences to crash.
So I went back to you way of implementing the ribbon. I added custom functions to show/hide and autohide the ribbon.

I also added the menus from the standard menubar to the application button. 
![image](https://github.com/user-attachments/assets/3d66a9e4-a613-44d3-ad3b-dee8652c8373)

With the "alt" key you can show/hide the standard menubar.

what do you think of my approach so far? I think most issues are handled without changes needed for all workbenches.
On the forum they are going into another direction, but I quite like the use of the ribbon, so maybe we can release it as an add-on?

kind regards

APEbbers committed:
Fixed issue with selecting an workbench that is not in the Json file

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #42 from APEbbers/Develop

Fixed issue with selecting a workbench that is not in the Json file

geolta commented:
> I don’t think it is possible to split up the pull request. But maybe you can create a second branch? I then create a pull request to that one, so you can review it without changing the main branch?

I'm not sure, whether I fully understand your suggestion. What would be the advantage of a second branch?


Very nice that the menu bar is now squashed into the FreeCAD icon!

I think, it definitely makes sense to publish this as an addon for the Addon Manager. If feasible, it would be great if we could manage to do this before the 1.0 release (sorry again for my late replies).
Do you have experience in bringing stuff to the Addon Manager?

I will have a detailed look at your changes tomorrow.
Feature wise I think they are quite neat :)

APEbbers committed:
Added scroll functionality for when the ribbon is wider than the screen

APEbbers committed:
Fixed issue whre custompanel was not properly removed from json file

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #43 from APEbbers/Develop

Develop

APEbbers committed:
changed te menu names to be more in line with FreeCAD native

APEbbers commented:
You right, you can also test my fork directly.

I've created 2 addons previously, so releasing it to the addon manager is no problem. I'll create a wiki page first, so that there is a manual in place? 

btw:
I added a dependency to the package.xml: keyboard. This is needed for detecting the alt key press.
I've also implemented a scroll function for when the ribbon has more panels than that fit on your screen

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #44 from APEbbers/Develop

changed te menu names to be more in line with FreeCAD native

geolta commented:
Hi, I have tested most of your changes. Here are my remarks :)

- Installation: The Addon Manager cannot install the `pyqtribbon` and `keyboard` libraries, at leat not on the AppImage for linux (but it can install `pyqt` for whatever reason). I had to manually copy the libraries into the project folder. If I remember correctly, this also was the reason to include `pyqtribbon` directly.
- The keyboard library seem not to work on Wayland under linux (tested on Gnome 46). It says, I have to be the admin to execute `on_press_key`.
- The help buttons in Design dialog and in the ribbon bar seem not to work. This might also be some permission issue.
- It took me some time to figure out what the pin button of the ribbon does – adding a tooltip for this would be good
- Making every first button of a toolbar large, seems not to be the best idea to me. This leads to many strange designs like the point in Sketcher being large. I understand, that this is just a starting point, but I think, keeping everything small at the beginning would be better. But of course, in the end, it's a matter of taste.
- Adding a warning like "may take some time" to the "Open Design" button might be good
- Add hint to Design dialog that a restart is required for the changes to take effect.
- I don't know from my head how to get icons from FreeCAD instead of having them "hard coded", but I think that this might be better to also be able to automatically apply icon themes, etc.
- Is there a reason you have changed every name, author, maintainer **everywhere** across the entire project to you? I understand this for modules written by you, but we are still relying on work from HakanSeven and me.
- The README should get an update after all of these changes.

Best regards :)

APEbbers commented:
Hi thanks for testing! 

I've written my response per comment below:

> "Installation: The Addon Manager cannot install the pyqtribbon and keyboard libraries, at least not on the AppImage for linux (but it can install pyqt for whatever reason). I had to manually copy the libraries into the project folder. If I remember correctly, this also was the reason to include pyqtribbon directly."

I wasn't aware of the issues with AppImage. I'll look into that. Maybe also a right issue? We can always include the pyqtribbon again


> "The keyboard library seem not to work on Wayland under linux (tested on Gnome 46). It says, I have to be the admin to execute on_press_key.
The help buttons in Design dialog and in the ribbon bar seem not to work. This might also be some permission issue."

I've used packages that are standard included in python, but I can look into different packages. I'm working mostly on windows, but I have a chromebox with debian on it.


> "It took me some time to figure out what the pin button of the ribbon does – adding a tooltip for this would be good"

agreed, I'll add a tooltip for it

> "Making every first button of a toolbar large, seems not to be the best idea to me. This leads to many strange designs like the point in Sketcher being large. I understand, that this is just a starting point, but I think, keeping everything small at the beginning would be better. But of course, in the end, it's a matter of taste."

The script allows to add already customized workbenches. I'm thinking to add customization for internal workbenches as standard like the sketcher WB. For every external workbench we have than the first icon of every toolbar set to large.
After that, the user can customize the script as desired. (or add extra scripts to the folder)

> "Adding a warning like "may take some time" to the "Open Design" button might be good"

agreed. I'll add a warning

> "Add hint to Design dialog that a restart is required for the changes to take effect."

agreed. I add the same or similar dialog as used with the add-on manager

> "I don't know from my head how to get icons from FreeCAD instead of having them "hard coded", but I think that this might be better to also be able to automatically apply icon themes, etc."

Actually, all icons are now retrieved from FreeCAD. But they are still written to the json file. This way we can add custom icon functionality easily at a later point. The way it works is that the design dialog retrieves the icons from freecad and writes the name to the json file. FCBinding.py, first retrieves the icon from FreeCAD. Then tries to get the icon from the Json file. If this succeeds, that icon is used. If it fails, the icon from freecad is used.

 

> "Is there a reason you have changed every name, author, maintainer everywhere across the entire project to you? I understand this for modules written by you, but we are still relying on work from HakanSeven and me."

i went a bit over board there. i noticed that the license in the files didn't match the MIT license, so I changed that and copied that to all files. I'll change the text to include both  your names as well? 

> "The README should get an update after all of these changes."

Agreed. This can be done when the wiki will be written.


kind regards

APEbbers committed:
Start implementation loading dialog
Updated licenses

APEbbers committed:
Updated loadingUI

APEbbers committed:
Set the maximum for the progressbar

APEbbers committed:
add a warning for long loading time
added python packages local for if not available

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
update_local import packages

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
added tooltip to pin button
different implementation of alt press

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

geolta commented:
Hi Paul,

> I wasn't aware of the issues with AppImage. I'll look into that. Maybe also a right issue? We can always include the pyqtribbon again

I think, this is not a rights issue, but rather a matter of which python executable is used and since the AppImage bundles its own python it doesn't access the packages installed system wide.

Maybe we can use git submodules to include pyqtribbon – but to be honest, I have never used them :)
  
> I've used packages that are standard included in python, but I can look into different packages. I'm working mostly on windows, but I have a chromebox with debian on it.

Maybe just add an if clause to the keyboard stuff to check that we're not running linux?

> The script allows to add already customized workbenches. I'm thinking to add customization for internal workbenches as standard like the sketcher WB. For every external workbench we have than the first icon of every toolbar set to large. After that, the user can customize the script as desired. (or add extra scripts to the folder)

You have convinced me :)

> > "I don't know from my head how to get icons from FreeCAD instead of having them "hard coded", but I think that this might be better to also be able to automatically apply icon themes, etc."
> 
> Actually, all icons are now retrieved from FreeCAD. But they are still written to the json file. This way we can add custom icon functionality easily at a later point. The way it works is that the design dialog retrieves the icons from freecad and writes the name to the json file. FCBinding.py, first retrieves the icon from FreeCAD. Then tries to get the icon from the Json file. If this succeeds, that icon is used. If it fails, the icon from freecad is used.

Ah, I meant the FreeCAD logo and things like the help button. Sorry for being imprecise.

> > "Is there a reason you have changed every name, author, maintainer everywhere across the entire project to you? I understand this for modules written by you, but we are still relying on work from HakanSeven and me."
> 
> i went a bit over board there. i noticed that the license in the files didn't match the MIT license, so I changed that and copied that to all files. I'll change the text to include both your names as well?

I have checked the license of HakanSevens repo again (https://github.com/HakanSeven12/Modern-UI?tab=GPL-3.0-1-ov-file#readme), it is a GPL 3.0 license so we're forced to use this as well. Maybe just use our three names? 

Best Regards! :)

APEbbers committed:
Started with keypress specific for linux, mac and windows.
linux/mac to be tested. Windows works.
Qt is a backup

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Added a restart dialog to standard_functions

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Fixed issue where iconsize was 0 in parameters

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
test another keypress method for linux

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Changed to pynput for keyboard listener
added a local version incase pynput is not installed

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
fixed wrong import local pynput

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
fixed wrong import local pynput

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

APEbbers committed:
temp

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Fixed copy error

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
removed pynput and switcht to Qt for keypress in linux

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Added restart dialog to desin and preference meny

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
implented the restart dialog including icons

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Removed prefixed icons in restard dialog

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Fixed issue where the separators where not sorted properly

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Update LICENSE

APEbbers committed:
Update LICENSE

APEbbers committed:
Changed the license to GPL-3.0 license and mentioned all involved

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #45 from APEbbers/Develop

Added restart dialog
removed keypress functions for linux
removed unused recources (icons)
fixed separators not sorted properly
updated licenses

APEbbers committed:
Update package.xml

APEbbers committed:
Merge pull request #46 from APEbbers/main

updates from main

APEbbers committed:
Replaced the helpbutton on the ribbon with FreeCAD's own help command
Added an about button to the application menu/Ribbon design
Added a function to get the repro adress. This to avoid hardcoding references to the wiki/about/readme pages

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Fixed issue with too many arguments

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

APEbbers committed:
Fixed issue where getReproAdress did not return a string
Added a few ifs to the repro adress function
Used the getReproAdress function also for the help button on the design dialog

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Copy error

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Fixed issue with help adresses
Fixed issue were the Help Icon was missing from the design dialog

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Fixed issue with preferences not saved when pressing restart in the restart dialog

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #47 from APEbbers/Develop

Develop

APEbbers committed:
removed temporary import of pyqtribbon

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #48 from APEbbers/Develop

Removed temporary import

APEbbers committed:
Set minimum height to avoid ribbon from dissapearing

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

APEbbers committed:
Merge pull request #49 from APEbbers/Develop

Develop

APEbbers commented:
Hi,

I fixed most of the issues:
- I added the pyqtribbon package as a backup. If this package is not installed, a local version will be used.
- I added a restart dialog to the design and preference dialogs. (freecad should now restart when clicked "restart")
- The keyboard library is now only imported when on windows or mac. (For linux I didn't find an alternative yet) 
   - For linux I used the Qt version for now. Only you have to click on the ribbon first to activate the function which is not ideal
- I added the tooltip to the pin button
- I hadn't problems with opening the help pages on linux. However:
   - I replaced the help button on the ribbon, with the one from FreeCAD itself
   - In the design dialog, the icon from the FreeCAD help is used.
   - I added an "About FreeCAD Ribbon" to the application menu->Customize
   - Both the help button in the design dialog and the about button refer to the repositories from which the ribbon is installed.
   - The wikipage on my repository was disabled. This is enabled now.
- I updated the license in both the files as in the repository 
 
I'll start with a draft design with the most important wb's (Part/PartDesign, TechDraw, Surface, BIM, FEM and CAM.
After that I'll update the default json file and the scripts.
Followed with the Wiki page and the readme

kind regards

APEbbers committed:
Fixed last commit

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #50 from APEbbers/Develop

Fixed last commit

APEbbers committed:
Fixed issue where separators where not placed corectly by pyqtribbon (bug!)

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

APEbbers committed:
Fixed previously solved issue where selecting a different workbench in the design tab raised an error, due to missing keys in the json file

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #51 from APEbbers/Develop

Workaround for bug in pyqtribbon with placing separators

APEbbers committed:
Fixed issue where the buttons were not sorted on action data

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #52 from APEbbers/Develop

Fixed sorting issue

APEbbers committed:
Added the restart function to the reset and restore buttons

APEbbers committed:
updated the json file

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #53 from APEbbers/Develop

Added the restart function to the reset and restore buttons
updated the RibbonStructure.json file

APEbbers committed:
Fixed issue were workbench not in Json file cause an error during loading

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #54 from APEbbers/Develop

Fixed issue were workbench not in Json file cause an error during loa…

APEbbers committed:
Updated the commandNames script. Now there will be an backup created and the output is RibbonStructure.Json instead of RibbonStructure_backup.jason as default

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #55 from APEbbers/Develop

Updated the commandNames script. Now there will be an backup created …

APEbbers committed:
Moved the about button one level up

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #56 from APEbbers/Develop

Develop

APEbbers committed:
Added images for readme and wiki

APEbbers committed:
Merge branch 'Develop' of https://github.com/APEbbers/FreeCAD-Ribbon into Develop

APEbbers committed:
added combined image

APEbbers committed:
Update README.md

APEbbers committed:
Update README.md

APEbbers committed:
Update README.md

APEbbers committed:
Update README.md

APEbbers committed:
Merge pull request #57 from APEbbers/Develop

Develop

APEbbers committed:
Fixed issue in Design dialog where workbench is not in json file causes error

APEbbers committed:
Merge pull request #58 from APEbbers/Develop

Fixed issue in Design dialog where workbench is not in json file caus…

APEbbers committed:
Fixed issue with sorted seperators (json file not up to date)

APEbbers committed:
Merge pull request #59 from APEbbers/Develop

Fixed issue with sorted seperators (json file not up to date)

APEbbers committed:
Fixed issue where the wrong menus where loaded on startup

APEbbers committed:
Merge pull request #60 from APEbbers/Develop

Fixed issue where the wrong menus where loaded on startup

APEbbers committed:
Fixed issue when removing a custom toolbar didn't removed the custom toolbar from the dropdown unit restart of the dialog

APEbbers committed:
test with stored list of toolbar to speedup the loading time

APEbbers committed:
add json to git ignore

APEbbers committed:
aborted test with toolbar list

APEbbers committed:
Merge pull request #61 from APEbbers/Develop

Develop

APEbbers committed:
schanged screenshot for readme

APEbbers committed:
Merge pull request #62 from APEbbers/Develop

schanged screenshot for readme

APEbbers committed:
Update README.md

APEbbers committed:
Changed adress for about page

APEbbers committed:
Merge pull request #63 from APEbbers/Develop

Changed adress for about page

APEbbers committed:
Started with input for the wikipage

geolta commented on README.md:
```diff
@@ -1,23 +1,43 @@
 ## Installation
 There are two options for installing this addon:
 ### Custom Repository for Addon Manager
-Go to `Edit/Preferences/Addon-Manager` and add the custom repository `https://github.com/geolta/FreeCAD-Ribbon.git` with the branch `main`. Now you can go to the Addon Manager (`Tools/Addon-Manager`) and install "FreeCAD Ribbon" (maybe you have to update your local cache first). Now restart FreeCAD and you will see a ribbon interface :)
```
Please revert the change in the URL

APEbbers replied:
Can you change this after you merged all the changes to your repo? I prefer to have not to have links to another repo in my own repro

geolta commented on README.md:
```diff
@@ -1,23 +1,43 @@
+* Geolta (https://github.com/geolta/FreeCAD-Ribbon)
 
-![Screenshots of FreeCAD with the Ribbon UI](Screenshot.png)
+![](https://github.com/APEbbers/FreeCAD-Ribbon/blob/main/Resources/Images/Screenshot.png)
```
Please revert the change in the URL of the image

APEbbers replied:
Can you change this when you merged all the changes to your repo? I prefer to have not to have links to another repo when I'm using my own repo for install

geolta commented on README.md:
```diff
@@ -1,23 +1,43 @@
 1. Restart FreeCAD.
 1. When you restarted you don't see any toolbar.
-1. Create a macro.
+1. Download the file
```
download which file?

APEbbers replied:
Changed the text to: "Create a new macro (Macro menu -> macros...-> create macro)"

geolta commented on Resources/Images/Screenshot.svg:
Well, this image looks broken :D

APEbbers replied:
I'm no longer using, so I've removed it.

geolta commented on package.xml:
```diff
@@ -5,27 +5,29 @@
   <pythonmin>3.8.0</pythonmin>
 
-  <url type="website">https://github.com/geolta/FreeCAD-Ribbon/wiki</url>
+  <url type="website">https://github.com/APEbbers/FreeCAD-Ribbon/wiki</url>
```
Please keep the URL

APEbbers replied:
Can you change this when you merged all the changes to your repo? I prefer to have not to have links to another repo when I'm using my own repo for install

geolta commented on package.xml:
```diff
@@ -5,27 +5,29 @@
+  <url type="website">https://github.com/APEbbers/FreeCAD-Ribbon/wiki</url>
 
-  <url branch="main" type="repository">https://github.com/geolta/FreeCAD-Ribbon</url>
+  <url branch="main" type="repository">https://github.com/APEbbers/FreeCAD-Ribbon</url>
```
Here as well

APEbbers replied:
Can you change this when you merged all the changes to your repo? I prefer to have not to have links to another repo when I'm using my own repo for install

geolta commented on package.xml:
```diff
@@ -5,27 +5,29 @@
-  <url type="readme">https://github.com/geolta/FreeCAD-Ribbon/blob/main/README.md</url>
-
-  <icon></icon>
+  <url type="readme">https://github.com/APEbbers/FreeCAD-Ribbon/blob/main/README.md</url>
```
and here

APEbbers replied:
Can you change this when you merged all the changes to your repo? I prefer to have not to have links to another repo when I'm using my own repo for install

geolta reviewed (commented):
Hi, very nice!
As far as I can see, there are only a couple of very small issues :)
If added them as reviews to the code.
Best regards!

geolta commented on .vscode/settings.json:
```diff
@@ -0,0 +1,13 @@
+  "python.formatting.provider": "none",
+  "editor.fontLigatures": false,
+  "workbench.colorCustomizations": {},
+  "autopep8.interpreter": ["D:\\Python\\Python3.10\\python.exe"],
```
This explicit path is not good, for other people (on other OS) this will be different. Please remove it and set this path as a user setting for you.

APEbbers replied:
I've removed the folder and updated .gitignore. It shouldn;t be in git, but just local for me

geolta commented on RibbonStructure.json.bak:
Does it make sense to add the backup file to git?

APEbbers replied:
indeed, old left over. I've removed it

geolta commented on Scripts/UpdateCommandNames.py:
```diff
@@ -0,0 +1,148 @@
+                            if value[2] != "" and value[2] != "...":
+                                Dict_RibbonCommandPanel["workbenches"][WorkBench][
+                                    "toolbars"
+                                ][ToolBar]["commands"][Command]["text"] = value[2]
```
This function has four nested for-loops plus three nested if clauses. That is quite a lot :D
Please break this into smaller functions or at least add a few comments.

APEbbers replied:
I've added some comments to clarify

geolta commented on Standard_Functions_RIbbon.py:
```diff
@@ -0,0 +1,380 @@
+
+    # # If UCase is true, convert to upper case
+    # if UCase is True:
+    #     Letter = Letter.upper()
```
Why is this as a comment? If it is not needed, please remove it or leave a comment, why it is there as a comment.

APEbbers replied:
I've a standard functions repo where this is a copy of. This is probably a left over from my other wb's. it is removed.

geolta commented on Standard_Functions_RIbbon.py:
```diff
@@ -0,0 +1,380 @@
+def GetNumberFromLetter(Letter):
+    # from openpyxl.utils import column_index_from_string
+
+    # Number = column_index_from_string(Letter)
```
here as well

APEbbers replied:
same as mentioned above

geolta reviewed (commented):
I have now also skimmed the code and found a few minor issues.

APEbbers committed:
update gitignore + removed old backup file

APEbbers committed:
Removed .vscode from git and updated .gitignore
Added commends to the UpdateCommandNames.py

APEbbers committed:
Merge pull request #64 from APEbbers/Develop

Develop

APEbbers committed:
Update README.md

APEbbers committed:
Changed issue where settings preferences where always saved despite cancel button clicked

APEbbers committed:
Merge pull request #65 from APEbbers/Develop

Changed issue where settings preferences where always saved despite c…

APEbbers committed:
Disable and hide the restore button when backup is disabled

APEbbers committed:
Fixed issue where Backup always was enabled

APEbbers committed:
Scripts and json files were missing the open button in the quickaccess toolbar

APEbbers committed:
Removed powerpoint files from the repo

APEbbers committed:
Merge pull request #66 from APEbbers/Develop

Develop

APEbbers committed:
updated the scripts

APEbbers committed:
Merge pull request #67 from APEbbers/Develop

updated the scripts

APEbbers commented:
Hi,

I've finished the Wiki page as well.  I believe the addon is now ready to be added to the addon manager.
at the bottom of [https://wiki.freecad.org/Add_Workbench_to_Addon_Manager](url) you can find how to add this to the add-on manager.

Regarding the links in the readme and package.xml, as mentioned I'd like to keep these links for my repo. If you change this after you have merged the pull request, that would be great. In the code itself you don't have to change the address. The code looks in the .git directory for the correct address.

You might need to copy my wiki content to yours. I don't believe this will be done automatically by the pull request. 
The helpbutton in the Ribbon Design menu is looking for a wiki page. If you don't want use a wiki page you have to change:  
- line 414 in FCBinding.py
- line 1628 in LoadDesign_Ribbon.py
  
kind regards,

Paul

APEbbers committed:
Add PieMenu as conflicting addon to package.xml

APEbbers committed:
Merge pull request #68 from APEbbers/Develop

Add PieMenu as conflicting addon to package.xml

APEbbers committed:
Update LoadDesign_Ribbon.py

APEbbers committed:
Removed PieMenu as conflict from package.xml
PieMenu is not compatible with the latest dev version of FreeCAD

APEbbers committed:
Merge pull request #69 from APEbbers/Develop

Removed PieMenu as conflict from package.xml

APEbbers committed:
Added function to add indivial views toolbar on older version of FreeCAD

APEbbers committed:
Fixed PySide

APEbbers committed:
Added function to add toolbar "Induvidual views" to older versions of freecad

APEbbers committed:
Merge pull request #70 from APEbbers/Develop

Added function to add indivial views toolbar on older version of FreeCAD

APEbbers committed:
Dropdowns are now possible for quick commands

APEbbers committed:
Fixed issue where the stylesheet was not read properly

APEbbers committed:
Ui Tweaks + update default stylesheet

APEbbers committed:
Fixed issue where settings where reset after selecting a different stylesheet

APEbbers committed:
UPdate to default css

APEbbers committed:
Merge pull request #71 from APEbbers/Develop

Futher improvements

APEbbers committed:
Added code to maipuilate the stylesheet. this to get the correct border color for an drop down arrow

APEbbers committed:
Fixed border color for buttons (hover)
Fixed black spot behind quickackess toolbar

APEbbers committed:
Merge pull request #72 from APEbbers/Develop

Futher tweaks with qss file

APEbbers committed:
Update README.md

APEbbers committed:
Update FCBinding.py

APEbbers committed:
Started with translations

APEbbers committed:
fixed type

APEbbers committed:
add missing translation

APEbbers committed:
Added translation file + script

APEbbers committed:
Merge branch 'main' into Develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

APEbbers committed:
Merge pull request #73 from APEbbers/Develop

Preparation for translations

geolta commented:
Nice, I will merge this now :)
